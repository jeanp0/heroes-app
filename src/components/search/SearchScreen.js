import React, { useMemo } from "react";
import { useForm } from "../../hooks/useForm";
import { useQuery } from "../../hooks/useQuery";
import { getHeroesByName } from "../../selectors/getHeroesByName";
import { HeroCard } from "../heroes/HeroCard";

export const SearchScreen = ({ history }) => {
  const query = useQuery();
  const q = query.get("q") || "";
  const [{ searchText }, handleInputChange] = useForm({ searchText: q });
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q]);

  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${searchText}`);
  };

  return (
    <div>
      <h1>Search</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form className="input-group" onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Find your hero"
              className="form-control"
              name="searchText"
              autoComplete="off"
              value={searchText}
              onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-outline-primary">
              Search
            </button>
          </form>
        </div>

        <div className="col-7">
          <h4>Results</h4>
          <hr />
          {!q && (
            <div className="alert alert-info animate__animated animate__fadeIn">
              Search a hero
            </div>
          )}
          {q && !heroesFiltered.length && (
            <div className="alert alert-danger animate__animated animate__fadeIn">
              There is no a hero with {q}
            </div>
          )}
          {heroesFiltered.map((hero) => (
            <div key={hero.id} className="animate__animated animate__fadeIn">
              <HeroCard {...hero} />
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};
