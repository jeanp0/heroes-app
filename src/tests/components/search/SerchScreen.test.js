import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("Pruebas en <SearchScreen />", () => {
  test("should mostrarse correctamente con valores por defecto", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search"]}>
        <Route paht="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Search a hero");
  });

  test("should mostrar a batman y el input con el valor del query string", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <Route paht="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper.find("input").prop("value")).toBe("batman");
    expect(wrapper).toMatchSnapshot();
  });

  test("should mostrar un error si no se encuentra el hero", () => {
    const q = "batmanasd";
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${q}`]}>
        <Route paht="/search" component={SearchScreen} />
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      `There is no a hero with ${q}`
    );
  });

  test("should de llamar el push del history", () => {
    const q = "batman";
    const history = { push: jest.fn() };
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?q=${q}`]}>
        <Route
          paht="/search"
          component={(props) => <SearchScreen history={history} />}
        />
      </MemoryRouter>
    );
    wrapper.find("input").simulate("change", {
      target: { name: "searchText", value: q },
    });

    wrapper.find("form").simulate("submit", { preventDefault() {} });
    expect(history.push).toHaveBeenCalledWith(`?q=${q}`);
  });
});
