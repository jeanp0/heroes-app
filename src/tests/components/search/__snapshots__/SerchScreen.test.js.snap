// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> should mostrar a batman y el input con el valor del query string 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="input-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="animate__animated animate__fadeIn"
        key="dc-batman"
      >
        <div
          className="card text-dark bg-light mb-3"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row g-0"
          >
            <div
              className="col-md-4 m-auto"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <div
                  className="d-grid"
                >
                  <a
                    className="btn btn-dark btn-sm"
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    MÃ¡s..
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> should mostrar un error si no se encuentra el hero 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="input-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batmanasd"
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger animate__animated animate__fadeIn"
      >
        There is no a hero with 
        batmanasd
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> should mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        className="input-group"
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
